### DISCUSSION TITLE
Questions about root offsets.....

### DISCUSSION THREAD

**Discussion by venumspyder** [jrouwe/JoltPhysics#918] [Q&A] [answered]
I've been going through the rag doll samples and have a question about root offsets. For example this code sample:

                           `// Create ragdoll
				Ref<Ragdoll> ragdoll = settings->CreateRagdoll(group_id++, 0, mPhysicsSystem);

				// Sample pose
				SkeletonPose pose;
				pose.SetSkeleton(settings->GetSkeleton());
				animation[random() % cAnimationCount]->Sample(0.0f, pose);

				// Override root
				**pose.SetRootOffset(start);**
				SkeletonPose::JointState &root = pose.GetJoint(0);
				**root.mTranslation = Vec3(0, cVerticalSeparation * (i + 1), 0);**
				root.mRotation = Quat::sRotation(Vec3::sAxisY(), angle(random)) * root.mRotation;
				pose.CalculateJointMatrices();

				// Drive to pose
				ragdoll->SetPose(pose);`

So you set the root offset in the pose which is then added to all the matrices when you set the pose of the rag doll. However, you also get the root joint state and set the local root translation and rotation which will then be set for the child joints. My question is why not do everything translation wise in the root joint state and not set the root offset in the pose? Or was this just added as a second option?

**Comment by jrouwe** [comment, marked as answer]
If you compile the library without `JPH_DOUBLE_PRECISION` then you could indeed put everything in the translation of the root joint. If you are using doubles, then you need a double precision translation vector to position the ragdoll, but the joints only use single precision translation (using doubles would be a waste for local transforms).

I could also have written it:

```
// Override root
pose.SetRootOffset(start + Vec3(0, cVerticalSeparation * (i + 1), 0));
SkeletonPose::JointState &root = pose.GetJoint(0);
root.mTranslation = Vec3::sZero();
root.mRotation = Quat::sRotation(Vec3::sAxisY(), angle(random)) * root.mRotation;
pose.CalculateJointMatrices();
```

without `JPH_DOUBLE_PRECISION` it could have been:

```
// Override root
SkeletonPose::JointState &root = pose.GetJoint(0);
root.mTranslation = Vec3(start + Vec3(0, cVerticalSeparation * (i + 1), 0));
root.mRotation = Quat::sRotation(Vec3::sAxisY(), angle(random)) * root.mRotation;
pose.CalculateJointMatrices();
```
